### PR TITLE
 Show function names in explore tool instead of only function indices

### PR DESCRIPTION
**Pull request by lpereira**
Another one in the "let's make debugging a bit easier": show function names in the "explore" tool.  (There's a bit of code repetition here to clean up the names, but it's slightly different when we're considering HTML output, so I think that's OK.)

Here's how it works on my machine:
image

### CODE REVIEW

**Comment by jameysharp** [comment]
Very nice!

This can be further improved by demangling the function names first. For some example code that does that, see: https://github.com/bytecodealliance/wasmtime/blob/54e53cc77a29c8e4eb819333d81ef4158685690a/crates/wasmtime/src/runtime/profiling.rs#L194-L200

I don't think we need to be nearly so careful about sanitizing function names in this setting. These strings are first serialized to JSON, and that should ensure that any string we provide is valid JSON. The JSON string is then used in a `<script>` tag; we might need to escape the JSON for embedding in HTML, but `<script>` tags are implicitly treated as CDATA in HTML5, right? Finally, the strings embedded in the JSON are spliced into the DOM using `textContent`, so they don't need to be escaped at that point.

As a result I think we should just give the user whatever they provided, preferably after demangling.

File: crates/wasmtime/src/runtime/module.rs
Comment (fitzgen) [inline crates/wasmtime/src/runtime/module.rs:1039]: We can avoid looking up the function index via instruction offset since `function_locations` should already return the functions in order:

```suggestion
        self.function_locations().enumerate().map(|(idx, (offset, len))| {
            let idx = u32::try_from(idx).unwrap();
```

**Review by fitzgen — COMMENTED** [review]
Thanks! One nitpick below before we merge this.

**Comment by fitzgen** [comment]
Agreed demangling would be great -- also fine with deferring that to a follow up PR.

**Comment by lpereira** [comment]
Pushed the requested changes. Current output here is now this one (notice the pop-up showing the unmangled name):

image

**Comment by jameysharp** [comment]
Ooh, the pop-up is a nice touch. There's a lot more that could be done with `wasmtime explore` eventually.

I'd be happy to review this next week, or just as happy if you want to approve it @fitzgen.

File: crates/explorer/src/lib.rs
Comment (fitzgen) [inline crates/explorer/src/lib.rs:194]: I think it would be better to let `demangled_name` be an `Option<String>` rather than making a clone of `name` and then have the JS choose the demangled name when available and fall back to the plain name.

File: crates/wasmtime/src/runtime/module.rs
Comment (fitzgen) [inline crates/wasmtime/src/runtime/module.rs:1047]: Instead of stringifying the name with the index here, and then forcing the consumer of this API in `crates/explore/src/lib.rs` to parse the name out from stringified name out, can we just give the caller the various components before stringification? That is, return an iterator that yields items of type

```rust
(
    u32,            // wasm function index
    Option<String>, // name, if available
    usize,          // text section offset
    usize,          // length of function in text section
)
```

I think we could also update the existing `function_locations` method, and perhaps rename it, rather than defining a whole new method, since that method exists to basically enable `wasmtime-explore`.

**Review by fitzgen — COMMENTED** [review]
Thanks! Really excited to get this, and your other PR, in `wasmtime explore`.

I have a few comments and suggestions below that we should address before merging this however. I'll also hold off on reviewing the other PR until after this one merges.

Let me know if anything I said isn't clear or something like that. Thanks again!

File: crates/wasmtime/src/runtime/module.rs
Comment (lpereira) [reply crates/wasmtime/src/runtime/module.rs:1047]: Alright.

File: crates/explorer/src/lib.rs
Comment (lpereira) [reply crates/explorer/src/lib.rs:194]: OK.

File: crates/explorer/src/index.js
Comment (lpereira) [reply crates/explorer/src/index.js:192]: Sure.  It was the auto-formatting tool that my editor uses.  I'll have to force-push to fix this easily.

File: crates/explorer/src/index.js
Comment (lpereira) [reply crates/explorer/src/index.js:105]: OK.

**Comment by lpereira** [comment]
Alright, force-pushed (sorry!) due to the formatting changes, but it should be a whole lot clearer now.  I'll rebase the other PR once this one lands.

File: crates/wasmtime/src/runtime/module.rs
Comment (fitzgen) [inline crates/wasmtime/src/runtime/module.rs:1036]: I think `function_loactions` is only used by this method, so we can we remove it or make it non-`pub`, as I suggested previously?

Also I like the named struct with named fields, that's indeed better than a large-ish tuple, but the docs here should match that now.

And since this is really more than just function locations and names (it also has an index, for example) maybe we should call this method `function_metadatas` and the struct `FunctionMetadata`? Feel free to take-or-leave this last one...

**Review by fitzgen — COMMENTED** [review]
Thanks! Looks just about ready to merge, couple tiny things still to go first.

File: crates/wasmtime/src/runtime/module.rs
Comment (lpereira) [reply crates/wasmtime/src/runtime/module.rs]: Good catch, let me clean this up.

File: crates/explorer/src/index.js
Comment (lpereira) [reply crates/explorer/src/index.js]: Oops!

**Comment by lpereira** [comment]
Alright, @fitzgen, pushed a new revision that should address your comments.

**Review by fitzgen — APPROVED** [review]
Looks great, thanks! And thanks for your patience with review :)

**Pull request merged**
